File: app/src/main/java/com/example/firabasefirstexperience/java/activity/SearchActivity.java
package com.example.firabasefirstexperience.java.activity;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageButton;
import android.widget.SearchView;

import com.example.firabasefirstexperience.R;
import com.example.firabasefirstexperience.java.adapter.UserAdapter;
import com.example.firabasefirstexperience.java.model.User;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class SearchActivity extends AppCompatActivity {

    private RecyclerView recView_SearchUsers;
    private FirebaseUser fbUser;
    private List<User> listUsers;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        ImageButton ibSerach_back = findViewById(R.id.ibSearch_back);
        ibSerach_back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        Toolbar toolBarSearch = findViewById(R.id.toolBarSearch);
        toolBarSearch.setTitle("");
        setSupportActionBar(toolBarSearch);

        recView_SearchUsers = findViewById(R.id.recView_SearchUsers);
        recView_SearchUsers.setLayoutManager(new LinearLayoutManager(SearchActivity.this, RecyclerView.VERTICAL, false));
        recView_SearchUsers.setHasFixedSize(true);

    }

    @Override
    protected void onStart() {
        super.onStart();

        DatabaseReference reference = FirebaseDatabase.getInstance().getReference("Users");
        fbUser = FirebaseAuth.getInstance().getCurrentUser();

        listUsers = new ArrayList<>();

        if (reference != null) {
            reference.addValueEventListener(new ValueEventListener() {
                @Override
                public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                    listUsers.clear();

                    for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                        User user = snapshot.getValue(User.class);

                        if (user != null && fbUser != null) {

                            if (!user.getId().equals(fbUser.getUid())) {
                                listUsers.add(user);
                            }
                        }
                    }
                    UserAdapter searchUsersAdapter = new UserAdapter(SearchActivity.this, listUsers);
                    recView_SearchUsers.setAdapter(searchUsersAdapter);
                }

                @Override
                public void onCancelled(@NonNull DatabaseError databaseError) {

                }
            });
        }
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);

        MenuItem item = menu.findItem(R.id.item_search);
        SearchView sv_Users = (SearchView) item.getActionView();
        sv_Users.setIconified(false);
        sv_Users.setQueryHint("Search");


        sv_Users.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {

                searchUsers(newText);
                return true;
            }
        });

        return super.onCreateOptionsMenu(menu);
    }


    private void searchUsers(String newText) {

        List<User> myList = new ArrayList<>();
        for (User user : listUsers) {
            if (user.getUserName().toLowerCase().contains(newText.toLowerCase())) {
                myList.add(user);
            }
        }

        UserAdapter adapterUserSearch = new UserAdapter(SearchActivity.this, myList);
        recView_SearchUsers.setAdapter(adapterUserSearch);
    }
}

File: app/src/main/java/com/example/firabasefirstexperience/kotlin/SearchActivity.kt
package com.example.firabasefirstexperience.kotlin

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.firabasefirstexperience.R
import com.example.firabasefirstexperience.java.adapter.UserAdapter
import com.example.firabasefirstexperience.java.model.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import kotlinx.android.synthetic.main.activity_search.*
import java.util.ArrayList

class SearchActivity : AppCompatActivity() {

    private var fbUser: FirebaseUser? = null
    private var usersList: List<User>? = null


    override fun onStart() {
        super.onStart()

        val reference: DatabaseReference = FirebaseDatabase.getInstance().getReference("Users")
        fbUser = FirebaseAuth.getInstance().currentUser

        usersList = ArrayList<User>()

        if(reference != null) {
            reference.addValueEventListener(object : ValueEventListener{
                override fun onCancelled(p0: DatabaseError) {

                }

                override fun onDataChange(p0: DataSnapshot) {
                    (usersList as ArrayList<User>).clear()

                    for (snapshot in p0.children) {
                        val user = snapshot.getValue(User::class.java)

                        if (user != null && fbUser != null) {

                            if (user.id != fbUser!!.uid) {
                                (usersList as ArrayList<User>).add(user)
                            }
                        }
                    }

                    val userAdapter = UserAdapter(this@SearchActivity, usersList)
                    recView_SearchUsers.adapter = userAdapter

                }

            })
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)

        setSupportActionBar(toolBarSearch)
        supportActionBar?.title = ""
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        recView_SearchUsers.layoutManager = LinearLayoutManager(this@SearchActivity, RecyclerView.VERTICAL, false)
        recView_SearchUsers.setHasFixedSize(true)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {

        val inflater: MenuInflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)

        val item: MenuItem = menu!!.findItem(R.id.item_search)
        val searchView: SearchView = item.actionView as SearchView
        searchView.isIconified = false
        searchView.queryHint = "Search"

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {

                searchView(newText)
                return true
            }
        })


        return super.onCreateOptionsMenu(menu)
    }

    private fun searchView(newText: String) {
        val myList = ArrayList<User>()

        for (user in usersList!!) {
            if (user.userName.toLowerCase().contains(newText.toLowerCase())) {
                myList.add(user)
            }
        }

        val userAdapter = UserAdapter(this@SearchActivity,myList)
        recView_SearchUsers.adapter = userAdapter
    }
}

